src/testing/mocking.md

```diff
@@ -0,0 +1,58 @@
+---
+minutes: 5
+---
+
+# Mocking
+
+If you want to do mocking, we recommend [Mockall]. You need to refactor your
```

Comment:
```suggestion
If you want to do mocking, AOSP recommends [Mockall], although there are several other crates available. You need to refactor your
```
..or something to that effect. I think the existing text implies "mockall is how people do test mocks in Rust", and that's not accurate.

Reply:
I did a survey and from what I can tell, Mockall is vastly more used than any other mocking library right now:

|                                               | Released | Updated | Dependants | Downloads/month |
|-----------------------------------------------|---------:|--------:|-----------:|----------------:|
| [Mockall](http://lib.rs/crates/mockall)       |     2019 |  _2023_ |        295 |              1M |
| [Mockers](http://lib.rs/crates/mockers)       |     2016 |  _2022_ |          3 |             560 |
| [Simulacrum](http://lib.rs/crates/simulacrum) |     2017 |    2018 |          1 |              6k |
| [Pseudo](http://lib.rs/crates/pseudo)         |     2017 |    2018 |          4 |             653 |
| [Double](http://lib.rs/crates/double)         |     2017 |    2019 |          1 |             208 |
| [Mockiato](http://lib.rs/crates/mockiato)     |     2019 |    2019 |          1 |              2k |
| [Mock-it](http://lib.rs/crates/mock-it)       |     2018 |  _2022_ |          2 |             406 |

Most libraries haven't been updated in a few years and most of them have very few reverse dependencies.

Of course Mockall isn't the end all of mocking, so a new library could be made which is even better. But right it's the one we will recommend to our engineers in Android.

At the end of the day, the course here is written for Android engineers. I've tried my best to take out the Android-specific parts and make as much as it generic. I think the advise here is also generic enough for this section.

However, let us add a note on the slide (and perhaps somewhere in the beginning) to say that the material here is an _opinionated_ guide to Rust. It tried to be fair and generic, but it is sponsored by Google, and it will recommend tools useful today to Android and Google engineers. Does that sound fair?

Reply:
Ah, thanks for that analysis of the libraries! A few search results suggested a much wider range of options. Given that information, this slide makes sense.

Reply:
Thanks! I will add a note explaining that the recommendations on the slide is current at of now and apply directly to Android. They might also apply to others, but as you point out, there _are_ other options out there!